src/authorize.ts

```diff
@@ -0,0 +1,175 @@
+
+  /**
+   * Called after the authorization process completed.
+   *
+   * Return false to indicate that the request was responded to, and that EveryAuth should not redirect the
+   * caller to the finishedUrl automatically.
+   */
+  onComplete?: (req: express.Request, res: express.Response, everyCtx: IEveryAuthContext) => Promise<boolean>;
+}
+
+/**
+ * @ignore
+ *
+ * Tries to determine the publically-visible URL for this part of the router.
+ */
+export const getHostedBaseUrl = (options: IEveryAuthOptions, req: express.Request): string => {
+  // Supplied in options; use that.
+  if (options.hostedBaseUrl) {
+    if (typeof options.hostedBaseUrl == 'string') {
+      return options.hostedBaseUrl;
+    } else if (typeof options.hostedBaseUrl == 'function') {
+      return options.hostedBaseUrl(req);
+    }
+  }
+
+  // Discovery time!
+  //
+  // Sometimes the originalUrl includes the entire request, sometimes it doesn't!
+  if (req.originalUrl.startsWith('http')) {
+    return req.originalUrl;
+  }
+
+  // req.hostname doesn't preserve the port, unfortunately, so test the Host header to see if it's present.
+  let port = '';
+  if (req.headers['host'] && req.headers['host'].includes(':')) {
+    port = `:${req.headers['host'].split(':')[1]}`;
+  }
+
+  // Return a hopefully valid URL.
+  return `${req.protocol}://${req.hostname}${port}${req.originalUrl}`;
+};
+
+/**
+ * Authorize a user with a particular service.
+ *
+ * @param serviceId The name of the service to authorize the user against.
+ * @param options Configuration options to control the behavior of this authorization.
+ * @return An Express Router
+ */
+export const authorize = (serviceId: string, options: IEveryAuthOptions): express.Router => {
+  const router = express.Router({ mergeParams: true });
+
+  const redirect = (req: express.Request, res: express.Response, ids: { userId: string; tenantId?: string }) => {
+    // Parse the finishedUrl, supporting both a plain path and a full url.
+    let finUrl: URL | undefined = undefined;
+    let fullUrl = false;
+
+    try {
+      finUrl = new URL(options.finishedUrl);
+      fullUrl = true;
+    } catch (_) {
+      // Not a fully qualified url, probably just a path.
+    }
+    if (!finUrl) {
+      finUrl = new URL(`http://localhost${options.finishedUrl}`);
+    }
+
+    finUrl.searchParams.append('userId', ids.userId);
+    if (ids.tenantId) {
+      finUrl.searchParams.append('tenantId', ids.tenantId);
+    }
+    if (req.query.error) {
+      finUrl.searchParams.append('error', req.query.error as string);
+    }
+
+    return res.redirect(fullUrl ? finUrl.toString() : `${finUrl.pathname}?${finUrl.searchParams.toString()}`);
+  };
+
+  router.get('/', async (req: express.Request, res: express.Response) => {
+    const userId = await options.mapToUserId(req);
+    const tenantId = options.mapToTenantId ? await options.mapToTenantId(req) : undefined;
```

Comment:
We are already opinionated about the semantics of fusebit.tenantId tag in @fusebit-int/framework. If we are opinionated in a different way in EveryAuth (by allowing a fusebit.userId tag with essentially the same semantics without fusebit.tenantId tag), it will make the "graduation" to Fusebit more complex.